cdap-ranger/cdap-ranger-lookup/src/main/java/co/cask/cdap/security/authorization/ranger/lookup/client/CDAPRangerLookupClient.java

```diff
@@ -0,0 +1,227 @@
+import org.apache.commons.logging.LogFactory;
+import org.apache.ranger.plugin.client.BaseClient;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.Nullable;
+
+/**
+ * CDAP Clients which to be used by ranger for listing cdap resources.
+ * Note: The log statements in this class are written in a formatted way to match the logging format
+ * used by other services in Apache Ranger since these log statements end up in ranger admin log file we
+ * want them to be in same format.
+ */
+public class CDAPRangerLookupClient {
+
+  private static final Log LOG = LogFactory.getLog(CDAPRangerLookupClient.class);
+  private static final long SERVICE_TIMEOUT_SECONDS = TimeUnit.MINUTES.toSeconds(30);
+
+  private static final String ERR_MSG = " You can still save the repository and start creating "
+    + "policies, but you would not be able to use autocomplete for "
+    + "resource names. Check xa_portal.log for more info.";
+  private AccessToken accessToken;
+  private final String serviceName;
+  private final String instanceURL;
+  private final String username;
+  private final String password;
+  private final NamespaceClient nsClient;
+
+  CDAPRangerLookupClient(String serviceName, String instanceURL, String username, String password) {
+    this.serviceName = serviceName;
+    this.instanceURL = instanceURL;
+    this.username = username;
+    this.password = password;
+    initConnection();
+    ClientConfig clientConfig = getClientConfig();
+    this.nsClient = new NamespaceClient(clientConfig);
+    //TODO create more cdap clients here
+  }
+
+  private void initConnection() {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("==> CDAPRangerLookupClient initConnection()");
+    }
+    if (username != null && password != null) {
+      // security is enabled, we need to get access token before checking system services
+      try {
+        LOG.info(String.format("Fetching access token with username %s and password ****", username));
+        accessToken = fetchAccessToken(username, password);
+      } catch (Exception ex) {
+        throw Throwables.propagate(ex);
+      }
+    }
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("<== CDAPRangerLookupClient initConnection()");
+    }
+  }
+
+  /**
+   * Tests that a connection can be established to CDAP with the given config. We do this by listing namespaces.
+   * @return Map<String, Object> Connection test response
+   */
+  Map<String, Object> testConnection() {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("==> CDAPRangerLookupClient testConnection()");
+    }
+    HashMap<String, Object> responseData = new HashMap<>();
+    List<String> testResult;
+    try {
+      testResult = getNamespaces(null);
```

Comment:
So there is two places where this will be used. First what you see right now testConnection and later in the lookup. The lookup will pass a empty or partial list of namespaces which will get modified. The testConnection passes null because for test connection we are not interested in looking what is in the list and hence there is no need to call contains. Since its or operation if the first statement implied to true the second is not evaluated hence we don't check contains (since it's a list and the contains time complexity is O(n) here)